### PR TITLE
FIX: if the user calls `matplotlib.use` disable backend fallback

### LINKED ISSUE #13883: In headless mode, matplotlib.use('tkagg') only errors after importing pyplot
### Bug report

**Bug summary**

Calling `matplotlib.use('tkagg')` doesn't work in headless mode. That's fine and to be expected.

What is bizarre is that this only results in an error when this is called *after* `matplotlib.pyplot` is imported. This results in a surprising and challenge to debug situation if `matplotlib.use('tkagg')` is called multiple times (e.g., in separate files).

**Code for reproduction**

Here's a script that I'm running with matplotlib 3.0.2:
```python
import os

# pretend to be in "headless" mode
os.environ.pop("DISPLAY", None)

# first module
import matplotlib
matplotlib.use('TkAgg')  # silently fails?
import matplotlib.pyplot

# in another imported module
import matplotlib
matplotlib.use('TkAgg')  # this is when matplotlib crashes!
import matplotlib.pyplot
```

**Actual outcome**

```
Traceback (most recent call last):
  File "/usr/local/google/home/shoyer/Downloads/use_tkagg_test.py", line 13, in <module>
    matplotlib.use('TkAgg')  # this is when matplotlib crashes!
  File "/usr/local/google/home/shoyer/miniconda3/envs/xarray-py37/lib/python3.7/site-packages/matplotlib/__init__.py", line 1391, in use
    switch_backend(name)
  File "/usr/local/google/home/shoyer/miniconda3/envs/xarray-py37/lib/python3.7/site-packages/matplotlib/pyplot.py", line 222, in switch_backend
    newbackend, required_framework, current_framework))
ImportError: Cannot load backend 'TkAgg' which requires the 'tk' interactive framework, as 'headless' is currently running
```

**Expected outcome**

It would be nice if matplotlib did any of these:
1. Gave a clear error message when `matplotlib.use()` was called the first time.
2. Gave a clear error message when `matplotlib.pyplot` is imported and an invalid backend is selected.
2. Silently allowed setting an invalid backend with `matplotlib.use()` even after pyplot is imported.

**Matplotlib version**

  * Operating system:
  * Matplotlib version: 3.0.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): NA
  * Python version: 3.7
  * Jupyter version (if applicable):
  * Other libraries: 

As far as I can tell, the source of the problem is that matplotlib only actually validates backends if `matplotlib.pyplot` is imported:
https://github.com/matplotlib/matplotlib/blob/d9030efd40fe4300e775db22ba55845399f6668f/lib/matplotlib/__init__.py#L1281=L1283

It seems like one good way to avoid this would be to use a consistent codepath, whether or not `matplotlib.pyplot` has already been imported, e.g., by always calling `matplotlib.pyplot.switch_backend()`. Or maybe `matplotlib.use()` could do some basic validation, e.g., to verify that the backend is a valid choice.

### PR DESCRIPTION
**Pull request by tacaswell**
If `rcParmas['backend_fallback'] == True` and we find the user has
requested an inconsistent backend with the currently running event
loop we resest `rcParams['backend']` to the auto backend selection
sentinel.  Immediately afterwards we call `switch_backend`.

In the case where we are on a headless linux machine we always detect
the running event loop as 'headless'.  This means that if any
interactive backend was selected (via an rcparams file, directly
setting `rcParams['backend']` or using `mpl.use`) on import of
`pyplot` we would, by default, run through the fallback logic and
select 'agg' as the backend.

This change records the user's intent to select a backend via `use` or
via setting an environmental variable by setting
`rcParam['backend_fallback'] = False` to prevent this fallback in the
case where the user has explicitly selected a backend in their code.
If the backend is selected an rcparam file, or directly setting the
value in the rcParam dict we will fallback.

closes #13883

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/__init__.py [hunk 1]
```diff
@@ -1301,10 +1301,12 @@ def use(backend, warn=False, force=True):
         # won't change the backend.  This is a bit of overkill as 'backend'
         # is already in style.core.STYLE_BLACKLIST, but better to be safe.
         rcParams['backend'] = rcParamsDefault['backend'] = name
-
+        rcParams['backend_fallback'] = \
+            rcParamsDefault['backend_fallback'] = False
 
 if os.environ.get('MPLBACKEND'):
     rcParams['backend'] = os.environ.get('MPLBACKEND')
+    RcParams['backend_fallback'] = False
```
Comment (anntzer) [inline lib/matplotlib/__init__.py]: lowercase?

**Comment by anntzer** [comment]
It's not clear to me that we want assignment to rcParams and use() have different semantics, but heh (we could use this patch in all cases and internally use `dict.__setitem__(rcParams, "backend", ...)` to not trigger this patch when we don't want to).

File: lib/matplotlib/__init__.py (hunk 1 above)
Comment (tacaswell) [reply lib/matplotlib/__init__.py]: yup...

**Comment by tacaswell** [comment]
> It's not clear to me that we want assignment to rcParams and use() have different semantics,

I agree, but was being lazy.... I guess the only place we would update this where we want to allow fallback is when loading a configuration file and it should be easy enough to handle that case.

**Comment by anntzer** [comment]
I haven't really spent time trying to untangle the logic in this PR, but I think it may make sense to kill (well, deprecate yada yada) the backend_fallback rcParam and have the behavior be
- if the backend is not specified in the matplotlibrc nor by matplotlib.use() nor by MPLBACKEND, go through the fallback logic
- else, use the requested backend.

**Comment by tacaswell** [comment]
> if the backend is not specified in the matplotlibrc

This is the case where people want this the most (when the system matplotlibrc is set up for an interactive session and you land on a headless server).

**Comment by anntzer** [comment]
Please document exactly how the different ways of setting the backend and the backend_fallback rcParam interact with one another, then.

File: lib/matplotlib/__init__.py [hunk 2]
```diff
@@ -954,7 +970,10 @@ def _rc_params_in_file(fname, fail_on_error=False):
 http://github.com/matplotlib/matplotlib/blob/master/matplotlibrc.template
 or from the matplotlib source distribution""" % (key, cnt, fname),
                   file=sys.stderr)
-
+    # setting the backend will force this to false, but we should respect
+    # the users input here so re-set it.
+    if try_backend_fallback_again is not None:
+        config['backend_fallback'] = try_backend_fallback_again
```
Comment (anntzer) [inline lib/matplotlib/__init__.py]: this logic is so ridiculously complicated :(

Comment (jklymak) [reply lib/matplotlib/__init__.py]: Yeah I don’t quite understand all the use cases we are trying to accommodate here.  But I’d sure be in favour of simpler logic and better error messages.

**Comment by tacaswell** [comment]
@dstansby due to concern about the complexity of this, I pushed this to 3.2.  Given that you labeled it release critical for 3.1 are you OK with this?

**Comment by dstansby** [comment]
Yep, no worries!

**Comment by tacaswell** [comment]
On the phone call @story645 suggested controlling if fallback is disabled by an additional kwarg to `mpl.use` which gets us out of the `use` and `__setitem__` behaving differentlly

```python
def use(arg, warn=False, force=False, * , allow_fallback=None):
    ...
    if allow_fallback is not None:
        rcParams['backend_fallback'] = allow_fallback
```

This seems much simpler and more explicit.

**Comment by anntzer** [comment]
What about having use() just always immediately set up the backend (i.e. it imports pyplot if it's not imported yet), instead of delaying it?  (Note that we can now switch between any backends even after pyplot is imported, as long as the event loop has not started.)

**Comment by jklymak** [comment]
Here is a case of a user being confused by the fallback:  https://stackoverflow.com/questions/55955817/cant-set-backend-in-matplotlib?noredirect=1#comment98565272_55955817

**Comment by tacaswell** [comment]
@jklymak This PR would fix that issue but it raises a different question if just checking for the `DISPLAY` env is sufficient to be sure that we are on a headless system.

**Comment by jklymak** [comment]
Yeah didn’t realize the OP was using ssh.  I’m not quite sure what checking DISPLAY gets you given that this does not even exist on windows and has a sketchy use on MacOS

**Comment by anntzer** [comment]
We only check for $DISPLAY on linux.

**Comment by tacaswell** [comment]
Closed in favor of #13977

**Pull request closed without merging**